Make InputBase use EventCallback for ValueChanged

SteveSandersonMS opened the pull request:
... so that the host component gets re-rendered automatically after each value change (as also happens when binding to DOM elements).

This is another fit-and-finish issue discovered by @rynowak during his app-building exercise.

SteveSandersonMS committed:
Make InputBase use EventCallback for ValueChanged

... so that the host component gets re-rendered automatically after each value change (like when binding to DOM elements).

SteveSandersonMS commented:
Again, I don't know whether it's still possible to include items like this for preview 3, but here's the PR anyway.

The value of including this is that is makes the new built-in input components (`InputText`, `InputSelect`, etc.) consistent with regular DOM elements in terms of what gets re-rendered after each edit. Without this PR, edits in the input components won't automatically re-render the container component, which is strange and can be very inconvenient.

davidfowl commented on src/Components/test/E2ETest/Tests/FormsTest.cs:
```diff
@@ -316,6 +316,27 @@ public void ValidationMessageDisplaysMessagesForField()
+
+            // Leaves previous value unchanged if new entry is unparseable
+            ticketClassInput.SelectByText("(select)");
+            await Task.Delay(500); // Not expecting any UI change, so wait a moment to see if one happens
```
No better way to test this?

SteveSandersonMS replied:
The inherent asynchrony of the server-side execution model makes it tricky for tests to assert "something shouldn't change". It's easy to assert when something should change within a given time span - that's what the `WaitAssert` calls do - but proving a negative is harder.

We *could* add some extra complications to the code under test for cases like this, whereby we try to trigger a secondary effect that we can wait for before running the "nothing changed" assertion. But it's a very unnatural way to write test code.

There are few cases where we really need "nothing changed" assertions in the Components E2E tests, so I haven't regarded it as a major issue. The same test cases also run automatically in the client-side (Blazor) runtime which is fully synchronous, so the assertion is guaranteed to be applied to the post-updates state anyway, meaning that we'd almost definitely catch any cases where the asynchony was hiding a bug.

If we later decide that we really need to solve this in a better way and not rely on waiting like this, I'd prefer to do it once in some central place, by having the test app somehow expose information about the number of renders, so the test code can wait until that number increments.

rynowak replied:
The E2E tests have a bunch of *wait until* loops and so on. 

In general waiting for someone not to happen is just problematic since it will fail to validate what you want if slow.

rynowak replied:
You could add validation and look for the `invalid` class. Not sure if that's a good fit or not for what's being tested.

rynowak replied:
I guess you have validation. So you could wait for the select to have `invalid` and then look at `selectedTicketClassDisplay`

SteveSandersonMS replied:
Great point. Done.

rynowak commented on src/Components/Components/src/Forms/InputComponents/InputBase.cs:
```diff
@@ -71,7 +71,7 @@ protected T CurrentValue
                 {
                     Value = value;
-                    ValueChanged?.Invoke(value);
+                    _ = ValueChanged.InvokeAsync(value);
```
Because this is fire and forget, it's possible for no one to see the error if one happens. 

I've considered in the past whether we should have a `HandleError(Task)` method on `ComponentBase` in case you find yourself in one of these situations. 

It doesn't seem in scope for this PR? but thoughts in general?

davidfowl replied:
Yes we should avoid unobserved tasks in general.

SteveSandersonMS replied:
Yes, I also agree we should explicitly handle this somehow, and the `ComponentBase.HandleError` approach sounds promising.

That's well out of scope for this PR, of course!

Also it's not a scenario that's going to affect conventional uses of `bind-*`, since with those, the "setter" code we generate is always synchronous anyway. You couldn't return a `Task` to the system even if you wanted to.

SteveSandersonMS committed:
Improve E2E test code

... so that the host component gets re-rendered automatically after each value change (like when binding to DOM elements).

Eilon reviewed (approved):
Approved for preview3.
